compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2020,13 +2021,29 @@ fn linker_with_args<'a>(
 
     // Link with the import library generated for any raw-dylib functions.
     for (raw_dylib_name, raw_dylib_imports) in
-        collate_raw_dylibs(sess, &codegen_results.crate_info.used_libraries)?
+        collate_raw_dylibs(sess, codegen_results.crate_info.used_libraries.iter())?
+    {
+        cmd.add_object(&archive_builder_builder.create_dll_import_lib(
+            sess,
+            &raw_dylib_name,
+            &raw_dylib_imports,
+            tmpdir,
+            true,
+        ));
+    }
+    // As with add_upstream_native_libraries, we need to add the upstream raw-dylib symbols in case
+    // they are used within inlined functions or instantiated generic functions. We do this *after*
+    // handling the raw-dylib symbols in the current crate to make sure that those are chosen first
+    // by the linker.
+    for (raw_dylib_name, raw_dylib_imports) in
+        collate_raw_dylibs(sess, codegen_results.crate_info.native_libraries.values().flatten())?
```

Comment:
Don't we get the import libraries when linking the rlibs already?

Reply:
If it is an rlib, then yes, but if it is a dylib then no - see the original issue that prompted this change: https://github.com/rust-lang/rust/pull/102327#issuecomment-1258863431

I considered filtering `native_libraries` to just those from dylib crates, but that could run into issues where there is a `raw-dylib` function in an rlib crate that is consumed by a dylib crate which is then consumed by the current crate.

Reply:
From the dependency_formats for the current crate type you could take the element corresponding to the originating crate and check that it is Linkage::Static. If not it is either a dylib itself or linked into a dylib.

Reply:
Ah, nice!